Applications/KeyboardMapper/main.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2020, Hüseyin Aslıtürk <asliturk@hotmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "KeyboardMapperWidget.h"
+#include <LibCore/ArgsParser.h>
+#include <LibGUI/AboutDialog.h>
+#include <LibGUI/Action.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/FilePicker.h>
+#include <LibGUI/Icon.h>
+#include <LibGUI/Menu.h>
+#include <LibGUI/MenuBar.h>
+
+int main(int argc, char** argv)
+{
+    const char* path = nullptr;
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(path, "Keyboard character mapping file.", "file", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    GUI::Application app(argc, argv);
+
+    auto app_icon = GUI::Icon::default_icon("app-keyboard-mapper");
+
+    auto window = GUI::Window::construct();
+    window->set_title("KeyboardMapper");
+    window->set_icon(app_icon.bitmap_for_size(16));
+    window->set_main_widget<KeyboardMapperWidget>();
+    window->resize(775, 315);
+    window->move_to(50, 50);
+    window->set_resizable(false);
+    window->show();
+
+    auto keyboard_mapper_widget = (KeyboardMapperWidget*)window->main_widget();
+    if (path != nullptr) {
+        keyboard_mapper_widget->load_from_file(path);
+    } else {
+        keyboard_mapper_widget->load_from_file("/res/keymaps/en.json");
+    }
+
+    // Actions
+    auto open_action = GUI::CommonActions::make_open_action(
+        [&](auto&) {
+            Optional<String> path = GUI::FilePicker::get_open_filepath("Open");
+            if (path.has_value()) {
+                keyboard_mapper_widget->load_from_file(path.value());
+            }
+        });
```

Comment:
Alignment

Reply:
Hi,
When i remove it, i get following error.

```
[0/1] Re-running CMake...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/asliturk/git-repos/serenity/Build
[1/1] Linking CXX executable Userland/keymap
FAILED: Userland/keymap 
: && /home/asliturk/git-repos/serenity/Toolchain/Local/bin/i686-pc-serenity-g++  -Wall -Wextra -Werror -std=c++2a -fdiagnostics-color=always -Os -Wno-sized-deallocation -fno-sized-deallocation -fno-exceptions -fno-rtti -Wno-address-of-packed-member -Wundef -Wcast-qual -Wwrite-strings -Wimplicit-fallthrough -Wno-nonnull-compare -Wno-deprecated-copy -Wno-expansion-to-defined -DDEBUG -DSANITIZE_PTRS  --sysroot /home/asliturk/git-repos/serenity/Build/Root Userland/CMakeFiles/keymap.dir/keymap.cpp.o  -o Userland/keymap -L/home/asliturk/git-repos/serenity/Build/Libraries/LibC   -L/home/asliturk/git-repos/serenity/Build/Libraries/LibM   -L/home/asliturk/git-repos/serenity/Build/Libraries/LibPthread -Wl,-rpath,/home/asliturk/git-repos/serenity/Build/Libraries/LibC:/home/asliturk/git-repos/serenity/Build/Libraries/LibM:/home/asliturk/git-repos/serenity/Build/Libraries/LibPthread:  Libraries/LibCore/libcore.a  Libraries/LibKeyboard/libkeyboard.a  Libraries/LibC/libc.a  Libraries/LibC/libcrt0.a && :
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: Libraries/LibKeyboard/libkeyboard.a(CharacterMapFile.cpp.o): in function `Keyboard::CharacterMapFile::load_from_file(AK::String const&)':
CharacterMapFile.cpp:(.text+0x3a9): undefined reference to `Core::File::File(AK::StringView const&, Core::Object*)'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x3d1): undefined reference to `Core::File::open(Core::IODevice::OpenMode)'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x41e): undefined reference to `Core::IODevice::error_string() const'
/home/asliturk/git-repos/serenity/Toolchain/Local/lib/gcc/i686-pc-serenity/10.1.0/../../../../i686-pc-serenity/bin/ld: CharacterMapFile.cpp:(.text+0x45e): undefined reference to `Core::IODevice::read_all()'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
 ~
```